### PR TITLE
[v7r1] Completing jobs vs Completed

### PR DESCRIPTION
**Pull request by phicharp**
This PR replaces the status during finalisation with `Completing`

BEGINRELEASENOTES

*WMS
CHANGE: As explained in https://github.com/DIRACGrid/DIRAC/issues/4412, `Completed` jobs is used as major status for jobs during completion while it is also used for jobs that have a pending request and require a further action before being `Done` or `Failed`.
CHANGE: `Completing` jobs are treated exactly like `Running` jobs, i.e. set Stalled if they didn't give sign of life, and are not set `Failed` directly as before. Jobs to be checked are those that had not sent a heartbeat within the `stalledTime` period, in order to reduce the number of jobs to check.
NEW: Created a file `JobStatus.py` for replacing the literal job statuses, and updated many modules using these new variables. Note that there are still many others to change, but could be done gradually...

ENDRELEASENOTES

### CODE REVIEW

File: WorkloadManagementSystem/DB/JobDB.py [hunk 1]
```diff
@@ -66,7 +66,7 @@
 #############################################################################
 
 JOB_STATES = ['Submitting', 'Received', 'Checking', 'Staging', 'Waiting', 'Matched',
-              'Running', 'Stalled', 'Done', 'Completed', 'Failed']
+              'Running', 'Stalled', 'Completing', 'Done', 'Completed', 'Failed']
 JOB_FINAL_STATES = ['Done', 'Completed', 'Failed']
```
Comment (fstagni) [inline WorkloadManagementSystem/DB/JobDB.py]: I notice now that `Completed` is both in `JOB_STATES` and in `JOB_FINAL_STATES`. Maybe there's (or there was) a good reason, does it still stand?

Comment (phicharp) [reply WorkloadManagementSystem/DB/JobDB.py]: Well, Completed was already in FINAL_STATES although it was not a final state for jobs (but now these are indeed in principle the only statuses a job should have when terminating at a site), but my impression is that this is irrelevant ;-). 
It is only used for counting in the SiteSummaryWeb, therefore I think we can leave it like this...

**Comment by phicharp** [comment]
Should I care about the unsuccessful checks although they seem unrelated with my PR?

**Comment by fstagni** [comment]
No. The `pytest` is due to a security library and sometimes it fails. The pep8 check failed because you pointed your PR initially to `integration` branch, and changed later. If you push a commit it will run correctly.

File: WorkloadManagementSystem/DB/JobDB.py (hunk 1 above)
Comment (fstagni) [reply WorkloadManagementSystem/DB/JobDB.py]: OK, leave it. 

Another thing: are updates to StalledJobAgent necessary or not?

**Comment by phicharp** [comment]
Concerning `StalledJobAgent`, it should indeed be slightly modified in order to consider jobs in status `Completing` i.e. when a job is killed while in this status it should be moved to Stalled. This will even be easier than currently as for `Completed` there was always the potentiality of an upload waiting for the destination SE to be unbanned, which probably was the reason for setting the deadline to 7 days... 
I'll look at this agent right now, thanks for reminding me...

**Comment by fstagni** [comment]
Are you done with this? In that case please remove the WIP in the title. And @atsareg please review.

**Comment by phicharp** [comment]
There is one point I don't understand at lines 303 to 308 of `StalledJobAgent`:
If the agent is disabled, the major status is not set but the minor status is set if defined... This doesn't sound logical...
At line 303 I would set `result = S_ERROR('Disabled mode')`

**Comment by phicharp** [comment]
I will again look at the logic of the `StalledJobAgent` if one can make it more compact: there is quite some duplicated code...

**Comment by atsareg** [comment]
Few other places where Completing should be added or replace Completed:
WorkloadManagementSystem/DB/JobDB.py
WorkloadManagementSystem/Service/JobMonitoringHandler.py|JobStateUpdateHandler.py
Core/Utilities/Graphs/Palette.py
RequestManagementSystem/Client/ReqClient.py
ResourceStatusSystem/Command/JobCommand.py|JobDoneRatioPolicy|JobEfficiencyPolicy and corresponding Test modules (also in comments)

In particular, Completed should be removed from

**Comment by phicharp** [comment]
@atsareg: I will look into those places...

**Comment by phicharp** [comment]
`JobDB.py`: I think the change had been done

`JobStateHandler.py`: IIUC there is no need for a change, the end time is set for `Completed`

`ReqClient.py`: OK as is since this is for pending requests

`JobEfficiencyPolicy.py`: should be even better now as `Completed` is really a final status

`Palette.py`: Yes, but which coloraturas should I use? I have no clue ;-)

**Comment by atsareg** [comment]
For Palette you can use the same color for Completing as for Completed I think.

**Comment by phicharp** [comment]
> For Palette you can use the same color for Completing as for Completed I think.

I used a coloraturas closer to `Running` as indeed while `Completing` the jobs is still technically running on the WN...

**Comment by phicharp** [comment]
Not clear to me why `InstallServer` fails in the tests...

**Comment by chrisburr** [comment]
It's a bug from #4415 that is affecting all of the CI runs. It's not clear (to me) if it's an issue with DIRAC or the test but me and Chris will discuss it on Monday (see #4417).

File: WorkloadManagementSystem/Agent/StalledJobAgent.py [hunk 2]
```diff
@@ -116,43 +111,51 @@ def execute(self):
...
         self.__updateJobStatus(job, 'Stalled')
         stalledCounter += 1
       else:
         self.log.verbose(result['Message'])
-        runningCounter += 1
+        aliveCounter += 1
 
-    self.log.info('Total jobs: %s, Stalled job count: %s, Running job count: %s' %
-                  (len(jobs), stalledCounter, runningCounter))
+    self.log.info('Total jobs: %d, Stalled jobs: %d, alive jobs: %d' %
```
Comment (fstagni) [inline WorkloadManagementSystem/Agent/StalledJobAgent.py]: For more clear message:
```suggestion
    self.log.info('Total jobs: %d, Stalled jobs: %d, Running+Completing jobs: %d' %
```

**Comment by phicharp** [comment]
Back to WIP while introducing global variables for job statuses on @chaen 's suggestion

File: WorkloadManagementSystem/Agent/StalledJobAgent.py (hunk 2 above)
Comment (phicharp) [reply WorkloadManagementSystem/Agent/StalledJobAgent.py]: OK, will do

**Comment by fstagni** [comment]
Would you also rebase now, please?

**Comment by fstagni** [comment]
Looks OK to me, but @atsareg  should give the final review.

**Pull request merged**